File: app/src/main/java/ru/melod1n/vk/common/TimeManager.java
package ru.melod1n.vk.common;


import android.text.format.DateUtils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.Timer;
import java.util.TimerTask;

public class TimeManager {

    private static int currentHour;
    private static int currentMinute;
    private static int currentSecond;

    private static ArrayList<OnHourChangeListener> onHourChangeListeners = new ArrayList<>();
    private static ArrayList<OnMinuteChangeListener> onMinuteChangeListeners = new ArrayList<>();
    private static ArrayList<OnSecondChangeListener> onSecondChangeListeners = new ArrayList<>();
    private static ArrayList<OnTimeChangeListener> onTimeChangeListeners = new ArrayList<>();

    static void init() {
        Calendar calendar = Calendar.getInstance();
        currentHour = calendar.get(Calendar.HOUR);
        currentMinute = calendar.get(Calendar.MINUTE);
        currentSecond = calendar.get(Calendar.SECOND);

        processThread.start();
    }

    private static Timer timer = new Timer();

    private static Thread processThread = new Thread(() -> timer.schedule(new TimerTask() {
        @Override
        public void run() {
            Date date = new Date(System.currentTimeMillis());

            String sHour = new SimpleDateFormat("HH", Locale.getDefault()).format(date);
            String sMinute = new SimpleDateFormat("mm", Locale.getDefault()).format(date);
            String sSecond = new SimpleDateFormat("ss", Locale.getDefault()).format(date);

            int hour = tryToParseInt(sHour);
            int minute = tryToParseInt(sMinute);
            int second = tryToParseInt(sSecond);

            if (currentHour != hour) {
                currentHour = hour;

                if (onHourChangeListeners != null) {
                    for (OnHourChangeListener onHourChangeListener : onHourChangeListeners)
                        onHourChangeListener.onHourChange(hour);
                }

                if (onTimeChangeListeners != null) {
                    for (OnTimeChangeListener onTimeChangeListener : onTimeChangeListeners)
                        onTimeChangeListener.onHourChange(hour);
                }
            }

            if (currentMinute != minute) {
                currentMinute = minute;

                if (onMinuteChangeListeners != null) {
                    for (OnMinuteChangeListener onMinuteChangeListener : onMinuteChangeListeners)
                        onMinuteChangeListener.onMinuteChange(minute);
                }

                if (onTimeChangeListeners != null) {
                    for (OnTimeChangeListener onTimeChangeListener : onTimeChangeListeners)
                        onTimeChangeListener.onMinuteChange(minute);
                }
            }

            if (currentSecond != second) {
                currentSecond = second;

                if (onSecondChangeListeners != null) {
                    for (OnSecondChangeListener onSecondChangeListener : onSecondChangeListeners)
                        onSecondChangeListener.onSecondChange(second);
                }

                if (onTimeChangeListeners != null) {
                    for (OnTimeChangeListener onTimeChangeListener : onTimeChangeListeners)
                        onTimeChangeListener.onSecondChange(second);
                }
            }

        }
    }, 0, DateUtils.SECOND_IN_MILLIS));

    private static int tryToParseInt(String string) {
        try {
            return Integer.parseInt(string);
        } catch (Exception e) {
            e.printStackTrace();
            return -1;
        }
    }

    public static int getCurrentHour() {
        return currentHour;
    }

    public static int getCurrentMinute() {
        return currentMinute;
    }

    public static int getCurrentSecond() {
        return currentSecond;
    }

    public static boolean isMorning() {
        return getCurrentHour() > 6 && getCurrentHour() < 12;
    }

    public static boolean isAfternoon() {
        return getCurrentHour() > 11 && getCurrentHour() < 17;
    }

    public static boolean isEvening() {
        return getCurrentHour() > 16 && getCurrentHour() < 23;
    }

    public static boolean isNight() {
        return getCurrentHour() == 23 || (getCurrentHour() < 6 && getCurrentHour() > -1);
    }

    public static void addOnHourChangeListener(OnHourChangeListener onHourChangeListeners) {
        TimeManager.onHourChangeListeners.add(onHourChangeListeners);
    }

    public static void removeOnHourChangeListener(OnHourChangeListener onHourChangeListener) {
        TimeManager.onHourChangeListeners.remove(onHourChangeListener);
    }

    public static void addOnMinuteChangeListener(OnMinuteChangeListener onMinuteChangeListener) {
        TimeManager.onMinuteChangeListeners.add(onMinuteChangeListener);
    }

    public static void removeOnMinuteChangeListener(OnMinuteChangeListener onMinuteChangeListener) {
        TimeManager.onMinuteChangeListeners.remove(onMinuteChangeListener);
    }

    public static void addOnSecondChangeListener(OnSecondChangeListener onSecondChangeListener) {
        TimeManager.onSecondChangeListeners.add(onSecondChangeListener);
    }

    public static void removeOnSecondChangeListener(OnSecondChangeListener onSecondChangeListener) {
        TimeManager.onSecondChangeListeners.remove(onSecondChangeListener);
    }

    public static void addOnTimeChangeListener(OnTimeChangeListener onTimeChangeListener) {
        TimeManager.onTimeChangeListeners.add(onTimeChangeListener);
    }

    public static void removeOnTimeChangeListener(OnTimeChangeListener onTimeChangeListener) {
        TimeManager.onTimeChangeListeners.remove(onTimeChangeListener);
    }

    public interface OnHourChangeListener {
        void onHourChange(int currentHour);
    }

    public interface OnMinuteChangeListener {
        void onMinuteChange(int currentMinute);
    }

    public interface OnSecondChangeListener {
        void onSecondChange(int currentSecond);
    }

    public interface OnTimeChangeListener {
        void onHourChange(int currentHour);

        void onMinuteChange(int currentMinute);

        void onSecondChange(int currentSecond);
    }
}

File: app/src/main/java/ru/melod1n/vk/common/TimeManager.kt
package ru.melod1n.vk.common

import android.content.Context
import android.content.IntentFilter
import android.text.format.DateUtils
import ru.melod1n.vk.receiver.MinuteReceiver
import java.text.SimpleDateFormat
import java.util.*

object TimeManager {
    var currentHour = 0
        private set
    var currentMinute = 0
        private set
    var currentSecond = 0
        private set

    private val onHourChangeListeners: ArrayList<OnHourChangeListener> = ArrayList()
    private val onMinuteChangeListeners: ArrayList<OnMinuteChangeListener> = ArrayList()
    private val onSecondChangeListeners: ArrayList<OnSecondChangeListener> = ArrayList()
    private val onTimeChangeListeners: ArrayList<OnTimeChangeListener> = ArrayList()

    fun init(context: Context) {
        context.registerReceiver(MinuteReceiver(), IntentFilter("android.intent.action.TIME_TICK"))

//        context.registerReceiver()

//        val calendar = Calendar.getInstance()
//        currentHour = calendar[Calendar.HOUR]
//        currentMinute = calendar[Calendar.MINUTE]
//        currentSecond = calendar[Calendar.SECOND]
//        processThread.start()
    }

    fun broadcastMinute() {
        for (onMinuteChangeListener in onMinuteChangeListeners) {
            onMinuteChangeListener.onMinuteChange(0)
        }
    }

    private val timer = Timer()
    private val processThread = Thread(Runnable {
        timer.schedule(object : TimerTask() {
            override fun run() {
                val date = Date(System.currentTimeMillis())
                val sHour = SimpleDateFormat("HH", Locale.getDefault()).format(date)
                val sMinute = SimpleDateFormat("mm", Locale.getDefault()).format(date)
                val sSecond = SimpleDateFormat("ss", Locale.getDefault()).format(date)
                val hour = tryToParseInt(sHour)
                val minute = tryToParseInt(sMinute)
                val second = tryToParseInt(sSecond)

                AppGlobal.handler.post {
                    if (currentHour != hour) {
                        currentHour = hour
                        for (onHourChangeListener in onHourChangeListeners) onHourChangeListener.onHourChange(hour)
                        for (onTimeChangeListener in onTimeChangeListeners) onTimeChangeListener.onHourChange(hour)
                    }
                    if (currentMinute != minute) {
                        currentMinute = minute
                        for (onMinuteChangeListener in onMinuteChangeListeners) onMinuteChangeListener.onMinuteChange(minute)
                        for (onTimeChangeListener in onTimeChangeListeners) onTimeChangeListener.onMinuteChange(minute)
                    }
                    if (currentSecond != second) {
                        currentSecond = second
                        for (onSecondChangeListener in onSecondChangeListeners) onSecondChangeListener.onSecondChange(second)
                        for (onTimeChangeListener in onTimeChangeListeners) onTimeChangeListener.onSecondChange(second)
                    }
                }
            }
        }, 0, DateUtils.SECOND_IN_MILLIS)
    })

    private fun tryToParseInt(string: String): Int {
        return try {
            string.toInt()
        } catch (e: Exception) {
            e.printStackTrace()
            -1
        }
    }

    val isMorning: Boolean
        get() = currentHour in 7..11

    val isAfternoon: Boolean
        get() = currentHour in 12..16

    val isEvening: Boolean
        get() = currentHour in 17..22

    val isNight: Boolean
        get() = currentHour == 23 || currentHour < 6 && currentHour > -1

    fun addOnHourChangeListener(onHourChangeListeners: OnHourChangeListener) {
        TimeManager.onHourChangeListeners.add(onHourChangeListeners)
    }

    fun removeOnHourChangeListener(onHourChangeListener: OnHourChangeListener?) {
        onHourChangeListeners.remove(onHourChangeListener)
    }

    fun addOnMinuteChangeListener(onMinuteChangeListener: OnMinuteChangeListener) {
        onMinuteChangeListeners.add(onMinuteChangeListener)
    }

    fun removeOnMinuteChangeListener(onMinuteChangeListener: OnMinuteChangeListener?) {
        onMinuteChangeListeners.remove(onMinuteChangeListener)
    }

    fun addOnSecondChangeListener(onSecondChangeListener: OnSecondChangeListener) {
        onSecondChangeListeners.add(onSecondChangeListener)
    }

    fun removeOnSecondChangeListener(onSecondChangeListener: OnSecondChangeListener?) {
        onSecondChangeListeners.remove(onSecondChangeListener)
    }

    fun addOnTimeChangeListener(onTimeChangeListener: OnTimeChangeListener) {
        onTimeChangeListeners.add(onTimeChangeListener)
    }

    fun removeOnTimeChangeListener(onTimeChangeListener: OnTimeChangeListener?) {
        onTimeChangeListeners.remove(onTimeChangeListener)
    }

    interface OnHourChangeListener {
        fun onHourChange(currentHour: Int)
    }

    interface OnMinuteChangeListener {
        fun onMinuteChange(currentMinute: Int)
    }

    interface OnSecondChangeListener {
        fun onSecondChange(currentSecond: Int)
    }

    interface OnTimeChangeListener {
        fun onHourChange(currentHour: Int)
        fun onMinuteChange(currentMinute: Int)
        fun onSecondChange(currentSecond: Int)
    }
}